### ISSUE TITLE
[BUG] New pending block related issues

### ISSUE THREAD

**Issue by Bal7hazar** [dojoengine/dojo#1893]
**Describe the bug**

I just did some tests with the current dojo main(a3587b3) working with a Juno RPC on Sepolia L2 world (deployed with sozo), it works pretty well regarding indexing the pending blocks, just have some remaining issues sometimes but only fixed by restarting torii.

Syncing seems to be stoped du to the following error:
```go
2024-04-27T08:51:58.768902Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x69a0a511bf60f7f7337fbf49330b958bd750f2bba2e7f3f4fff60148baa1d93
2024-04-27T08:52:05.454569Z ERROR tori_core::engine: Syncing to head. error=Event without block number.
2024-04-27T08:52:07.606520Z  INFO tori_core::engine: Processed block. block_number=62277
```

Something mb linked to a ws connection which led to stop synching after the block finality (no idea how it could be related)
```go
2024-04-27T08:26:53.061114Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x751122493b59acd8c287cb5ca6b98933da941bb6cb00e2c0240b3f526a82a5
2024-04-27T08:26:53.102505Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x5326c293fb1c7aaca713a680274714613a4613d4980bfaa520bc246ac5aa080
2024-04-27T08:33:20.530867Z ERROR hyper::server::tcp: accept error: Too many open files (os error 24)
2024-04-27T08:33:20.531030Z ERROR torii_server::proxy: HyperError(hyper::Error(Io, Os { code: 54, kind: ConnectionReset, message: "Connection reset by peer" }))
2024-04-27T08:33:23.055796Z  INFO tori_core::engine: Processed block. block_number=62274
```

Something mb linked to the rpc, but restarting torii fixed the issue, so Im not sure (kept trying until I stoped)
```go
2024-04-27T08:39:05.626465Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x6e2352ea0521b6ae968c8e362ef8ccda536971885ceabc3655aa4b1fc3729ab
2024-04-27T08:39:07.741936Z ERROR tori_core::engine: Syncing to head. error=error sending request for url (https://rpc.nethermind.io/sepolia-juno/?apikey={API_KEY}): error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known
2024-04-27T08:39:09.747948Z ERROR tori_core::engine: Syncing to head. error=error sending request for url (https://rpc.nethermind.io/sepolia-juno/?apikey={API_KEY}): error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known
2024-04-27T08:39:12.753550Z ERROR tori_core::engine: Syncing to head. error=error sending request for url (https://rpc.nethermind.io/sepolia-juno/?apikey={API_KEY}): error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known
[...]
```

Some random errors due to hash
```go
2024-04-27T08:11:25.500300Z  INFO tori_core::engine: Processed pending transaction. block_number=62271 transaction_hash=1126356539500894896308565386773833074793817076872509247551031353978698615907
2024-04-27T08:11:26.636584Z ERROR tori_core::engine: Getting block. error=TransactionHashNotFound
2024-04-27T08:11:28.779609Z  INFO tori_core::engine: Processed pending transaction. block_number=62271 transaction_hash=445778682905951849140979879339574963743084709757155500485972618910687136764
```

Same but turned in warning sometimes
```go
2024-04-27T08:24:24.743094Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x1f666faf41adb68a180e3382c58482a661ce6297943a622310f470738443187
2024-04-27T08:24:30.382552Z  WARN tori_core::engine: Processing pending transaction. error=TransactionHashNotFound transaction_hash=0x17a5ff51db7fbeda862cdc39032c8eb6c3468e9a6d133cd60c53e39acdf55e9
2024-04-27T08:24:31.523810Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x17a5ff51db7fbeda862cdc39032c8eb6c3468e9a6d133cd60c53e39acdf55e9
```

**To Reproduce**

I am play testing Paved(mainnet branch) with the following envs:
VITE_PUBLIC_SEPOLIA=true
VITE_PUBLIC_FEE_TOKEN_ADDRESS=0x044e6bcc627e6201ce09f781d1aae44ea4c21c2fdef299e34fce55bef2d02210
VITE_PUBLIC_MASTER_PRIVATE_KEY=0xaf7f0b88b7ee7faf8065517cd565d9f9dcd013dd2382a0f1e6b3559f0fe57a
VITE_PUBLIC_MASTER_ADDRESS=0x059b1a0c489b635d7c7f43594d187362ddd2dcea6c82db4eef2579fd185b3753
VITE_PUBLIC_NODE_URL=https://rpc.nethermind.io/sepolia-juno/?apikey={API_KEY}
VITE_PUBLIC_API_KEY={API_KEY}

If needed I can share my Juno API_KEY by dm.

**Additional context**

Sepolia L2

**Comment by glihm** [comment]
@Bal7hazar thanks for the report here. @Larkooo are we processing pending transaction by batch, or one by one?

**Comment by Bal7hazar** [comment]
Actually it seems that even with no error, the syncing is just stopping after the pending block is finalized:

```go
2024-04-27T15:41:38.828191Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x4b04c00b471ce3627abec4ba68644b2407e2098bec6cc17959737661557404d
2024-04-27T15:41:38.869309Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x2504089b444c0286d3bd74ca984cf8c31d3f3f2680160e21f321447596fb382
2024-04-27T15:41:38.905671Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x7a520762d6f4b181e8e9e7ac9330fc139cf090117c7908c44aa551eadc7df17
2024-04-27T15:41:51.232150Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x7c153d30f75bca4af82f283e7c9845fdc4c4c4cc5740023b95715553fd49eab
2024-04-27T15:41:59.742728Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x15502e66756c15aefdf42b1bd71ac0a2c6a0afa14db82f3edec6fee1c78a7a4
2024-04-27T15:41:59.907586Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x20564245d0cef8227cb78b4fff08efc2e374182d4626ff522a760ab6fa6537e
2024-04-27T15:42:00.063473Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x7fa8fc3f5f7b4fd0b12fef06ef7d6f7843c6bd8273db06622ca5965dd4baf20
2024-04-27T15:42:05.574423Z  INFO tori_core::engine: Processed block. block_number=62344
```

**Comment by Larkooo** [comment]
> @Bal7hazar thanks for the report here. @Larkooo are we processing pending transaction by batch, or one by one?

One by one. We need the txns receipts.
As for the errors outlined here. I think some of them are resolved, @Bal7hazar try pulling the latest changes from main. The "Getting block. TransactionHashNotFound" is fixed. Now it's just a warning. That's expected behaviour. It will retry until it gets it. An issue that might need addressing though is the Event without a transaction hash. Will take a look

**Comment by Bal7hazar** [comment]
>  I think some of them are resolved, @Bal7hazar try pulling the latest changes from main
The fixes were not part of https://github.com/dojoengine/dojo/pull/1798?
Since the PR has been merged I assumed I was testing your latest changes, but I could be wrong.

If it matters, the worst case is actually this one: https://github.com/dojoengine/dojo/issues/1893#issuecomment-2080945893

**Comment by Larkooo** [comment]
> > I think some of them are resolved, @Bal7hazar try pulling the latest changes from main
> 
> The fixes were not part of #1798? Since the PR has been merged I assumed I was testing your latest changes, but I could be wrong.
> 
> > The "Getting block. TransactionHashNotFound" is fixed. Now it's just a warning.
> 
> Indeed it is most of the time but sometimes it seems to be still reported as an Error
> 
> If it matters, the worst case is actually this one: [#1893 (comment)](https://github.com/dojoengine/dojo/issues/1893#issuecomment-2080945893)

From taking a look at your error logs, it seems like the TransactionHashNotFound error was indeed caused because you weren't using the most up to date version of torii. Your transaction hashes are formatted in decimal, which was the case when that issue was not addressed. The remaining issue that seems quite important is the event without block number and I just published a fix that should cover most cases where that happens, even though it should never happen theoretically.  I think for your other issues/errors, they might be linked to you getting rate limited by nethermind. I will do further testing next week.

**Comment by Bal7hazar** [comment]
Tests with the 1897

## Pending block sync stopped after block finalization (none error raised, just stop silently to index pending txs)

```
2024-04-27T20:07:24.639327Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x4a1b2d247f530dd9e6c5d0004f1a000e964439b1ee04742e73691a93cc554f9
2024-04-27T20:07:24.714939Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x1ae2cce9e3dff19fab7184e341632b517db7aeeed4622b573af1af959c6dc67
2024-04-27T20:07:24.810226Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x3fa6cee6bec26499a468460eadde231bf83ff9e3495440f28f9ba112ed96c31
2024-04-27T20:07:29.490970Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x318df8ee0e59570e975a4a38d68357c5c2560e0d68975613ff3726e9aeebf48
2024-04-27T20:07:29.531414Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x645836238271661722d4e18bfeb08a62fdd9245b57b5c86ee72d632b8790c17
2024-04-27T20:07:36.386307Z  INFO tori_core::engine: Processed block. block_number=62387
```

Note: related to https://github.com/dojoengine/dojo/issues/1893#issuecomment-2080945893

## Invalid transaction state

An issue just before the block has finalized

```go
2024-04-27T20:00:55.410808Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x797f96c787d7a74aeeb95e688d1631c78c526fffaa1677ef791e031fe1bdc0a
2024-04-27T20:00:55.450888Z  INFO tori_core::engine: Processed pending transaction. transaction_hash=0x5ae4265591a46016c69e94b1bc39dae29184344b3aece61d61164afacad0b0d
2024-04-27T20:01:03.269919Z ERROR tori_core::engine: Invalid transaction state. Could not fetch block number for event. transaction_hash=0x4d240b5196308e89c790ef5c8da0c0f67388c12e853ce19b2a8094d6e30b886
2024-04-27T20:01:03.269985Z ERROR tori_core::engine: Syncing to head. error=Expected finalized transaction receipt for event.
2024-04-27T20:01:05.411521Z  INFO tori_core::engine: Processed block. block_number=62386
```

[Link](https://sepolia.voyager.online/tx/0x4d240b5196308e89c790ef5c8da0c0f67388c12e853ce19b2a8094d6e30b886#events) to tx making the event failure (nothing wrong from first look in voyager)
